src/components/validation/runner/mod.rs

```diff
@@ -566,23 +583,25 @@ fn spawn_output_driver(
 
             // Update the runner metrics for the received event. This will later
             // be used in the Validators, as the "expected" case.
-            let mut output_runner_metrics = output_runner_metrics.lock().unwrap();
+            let mut output_runner_metrics = output_runner_metrics.lock().await;
 
             for output_event in events {
-                output_runner_metrics.received_events_total += 1;
                 output_runner_metrics.received_event_bytes_total += vec![output_event.clone()]
                     .estimated_json_encoded_size_of()
                     .get()
                     as u64;
 
-                if let Some(encoder) = maybe_encoder.as_ref() {
-                    let mut buffer = BytesMut::new();
-                    encoder
-                        .clone()
-                        .encode(output_event, &mut buffer)
-                        .expect("should not fail to encode output event");
-
-                    output_runner_metrics.received_bytes_total += buffer.len() as u64;
+                if !skip_output_driver_metrics {
+                    if let Some(encoder) = maybe_encoder.as_ref() {
+                        let mut buffer = BytesMut::new();
+                        encoder
+                            .clone()
+                            .encode(output_event, &mut buffer)
+                            .expect("should not fail to encode output event");
+
+                        output_runner_metrics.received_events_total += 1;
+                        output_runner_metrics.received_bytes_total += buffer.len() as u64;
+                    }
                 }
```

Comment:
It seems like if we're not skipping output driver metrics, then that would imply we have a sink component under test, which should be using a "real" output external resource... and that this logic should exist in there, similar to how you've changed things for collecting source metrics in input external resources.

I think it's way too confusing to have this here, because it means every external resource is having to pass back a reciprocal encoder to re-encode the event and track the encoded size instead of just... counting the actual bytes the external resource received in the first place.

Reply:
Yes think this was just a temporary thing I added so that I could test the sink validators with the external resource setting the expected values on one end, without changing it for the others yet. But I think the final solution must have the external resources setting the expected values and not have this boolean flag at all.

Reply:
> It seems like if we're not skipping output driver metrics, then that would imply we have a sink component under test, which should be using a "real" output external resource... and that this logic should exist in there, similar to how you've changed things for collecting source metrics in input external resources.

So, here if we are not skipping output driver metrics, it means it is a source or transform that is under test. This logic is needed because, in the source and transform case, we have a controlled edge (vector sink) getting the output and the output driver is the place where we have access to determining the expected values , but we don't want to do this when we are testing sinks, because in that case the external resource the sink is pushing to / pulling from, is where the expected values are available.